src/com/amazon/ion/BufferConfiguration.java

```diff
@@ -0,0 +1,186 @@
+         * enough to exceed the length of the longest top-level value plus any system values that precede it.
+         */
+        static final int DEFAULT_INITIAL_BUFFER_SIZE = 32 * 1024; // bytes
+
+        /**
+         * The initial size of the lookahead buffer, in bytes.
+         */
+        private int initialBufferSize = DEFAULT_INITIAL_BUFFER_SIZE;
+
+        /**
+         * The maximum number of bytes that will be buffered.
+         */
+        private int maximumBufferSize = Integer.MAX_VALUE;
+
+        /**
+         * The handler that will be notified when the maximum buffer size is exceeded.
+         */
+        private T eventHandler = null;
+
+        /**
+         * Sets the initial size of the buffer that will be used to hold the data between top-level values. Default:
+         * 64KB.
+         *
+         * @param initialBufferSizeInBytes the value.
+         * @return this Builder.
+         */
+        public final V withInitialBufferSize(final int initialBufferSizeInBytes) {
+            initialBufferSize = initialBufferSizeInBytes;
+            return (V) this;
+        }
+
+        /**
+         * @return the initial size of the lookahead buffer, in bytes.
+         */
+        public final int getInitialBufferSize() {
+            return initialBufferSize;
+        }
+
+        /**
+         * Sets the handler that will be notified when bytes are processed and when the maximum buffer size is exceeded
+         * (if applicable). If the maximum buffer size is finite (see {@link #withMaximumBufferSize(int)}, the handler
+         * is required to be non-null. Otherwise, the handler may be null, in which case the number of bytes processed
+         * will not be reported.
+         *
+         * @param handler the handler.
+         * @return this builder.
+         */
+        public final V withHandler(final T handler) {
+            eventHandler = handler;
+            return (V) this;
+        }
+
+        /**
+         * @return the handler that will be notified when the maximum buffer size is exceeded.
+         */
+        public final T getHandler() {
+            return eventHandler;
+        }
+
+        /**
+         * Set the maximum number of bytes between top-level values. This can be used to limit growth of the internal
+         * buffer. For binary Ion, the minimum value is 5 because all valid binary Ion data begins with a 4-byte Ion
+         * version marker and the smallest value is 1 byte. For delimited text Ion, the minimum value is 2 because the
+         * smallest text Ion value is 1 byte and the smallest delimiter is 1 byte. Default: Integer.MAX_VALUE.
+         *
+         * @param maximumBufferSizeInBytes the value.
+         * @return this builder.
+         */
+        public final V withMaximumBufferSize(final int maximumBufferSizeInBytes) {
+            maximumBufferSize = maximumBufferSizeInBytes;
```

Comment:
Should this throw an `IllegalArgumentException` if the buffer size is smaller than `getMinimumMaximumBufferSize()`?

EDIT: I see that this check happens in the `BufferConfiguration` constructor. Based on the constraints described in the javadoc I expected it to happen here. Either way seems ok.

Reply:
Yeah, since I couldn't do all of the validation related to `maximumBufferSize` here (e.g. comparing it to `initialBufferSize`, which might not yet be set), I decided to just do all of it in the constructor.